Pod html oo 20210512 02

jkeenan opened the pull request:
This pull request is the fifth of five, the result of which will be that Pod::Html will now internally be object-oriented.

jkeenan committed:
Introduce Testing::record_state_of_cache()

A method to assist in debugging cache problems.

Should assist in resolving https://github.com/Perl/perl5/issues/12271.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Reduce default arguments for Testing:_prepare_argstable()

Touch-up t/htmlview.t

jkeenan committed:
Start refactoring lib/Pod/Html.pm

Group all the "global" variables into a hashref $globals.  This variable
will be the return value of init_globals() within pod2html(), will be
passed through, and augmented by, parse_command_line() and then used
through the balance of pod2html().

In essence, so far this is just a renaming of variables.

Remove declarations of superseded variables.

Pass $globals to get_cache(), cache_key() and load_cache().

jkeenan committed:
Introduce Pod::Html::Auxiliary

This package will hold helper subroutines used within the main package
or in tests.  They can be placed in a separate module and imported into
Pod::Html because they won't depend on having the globals passed as an
argument.  They will also be potentially independently testable.

Start with html_escape().

Move anchorify(), htmlify() to Auxiliary.pm.  Also _unixify -- now as unixify().

Move relativize_url() to Auxiliary.

Move usage() to Auxiliary.

Move trim_leading_whitespace to Auxiliary.

Move parse_command_line() to Auxiliary.

Keep porting tests happy. Increment $VERSION.  Run: ./perl -Ilib
regen/lib_cleanup.pl

anchorify.t, eol.t:  Correct excessive corrections.

Standardize setting of $VERSION.

jkeenan committed:
Create sub write_file()

We now start placing parts of sub pod2html() into separate subs so that
we reduce the length of pod2html(), making it more readable and
(ultimately) preparing for method calls.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Remove last hard-tabs from Html.pm

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Create internal sub parse_input_for_podtree()

This clears up a bit of semantic confusion.  We were using one lexically
scoped variable -- my $parser -- to hold two different objects: one for
parsing input, one for writing output.  We can encapsulate the working
of the input parser to make the code more readable.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce internal sub refine_globals()

Encapsulate! Encapsulate!

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce internal sub set_Title();

Encapsulate more code within pod2html(). The return value is for
convenience.  It's the setting inside $globals that counts.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce generate_cache() and identify_input()

Further encapsulation of code to improve readability.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce refine_parser()

Further encapsulation of code internal to pod2html().

At this point we face obstacles:  %Pages; $output.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Re-arrange subs in calling order

I.e., first pod2html(), then its internal subs in order of (potential)
call.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Add TODO comments for $Podroot and %Pages

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Move some assignments back into Pod::Html

In the subroutine exported by Pod::Html::Auxiliary, little more should
be done than to wrap around GetOptions and provide a usage message if
needed.  This subroutine should not need to be aware of the globals
already under assignment in Pod::Html.

Hence, rename that subroutine to process_command_line(), limit what it
does, and have it return a reference to a hash.

In Pod::Html::pod2html(), introduce process_options() to handle the
initial merging of the data in $opts into $globals.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Start to make Pod::Html object-oriented.

For the time being, at least, we'll have pod2html() call a constructor.
All subs up through refine_globals have been turned into method calls.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
File::Spec::Unix is not needed within package Pod::Html

It's only needed within Pod::Simple::XHTML::LocalPodLinks, so let's
remove it from the first package.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Make get_cache(), cache_key() and load_cache() use object

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Have generate_cache() use object

Still have to deal with %Pages being outside the object.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Turn 3 more internal subroutines into method calls

identify_input(); parse_input_for_podtree(); set_Title_from_podtree();

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Convert 3 more subroutines to method calls

refine_parser(); feed_tree_to_parser(); write_file()

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Replace _save_pages() with _transform()

_save_pages() is a Pod::Simple::Search method which takes two specific
arguments.  From the point of view of making Pod-Html OO, it was
problematic because (a) it used a variable ($Podroot) defined outside of
its scope and (b) it assigned to a variable (%Pages) defined outside of
its scope.  It was therefore resistant to encapsulation.

Experimentation showed that if we use the return value of
Pod::Simple::Search::survey(), we could parse that hashref using what
was the guts of _save_pages() and assign explicitly to %Pages.

TODO:  Move %Pages within the Pod::Html object.  Handle $object
properly.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Move $output into the object

Correct dropped variable declaration.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce $self->{Pages}

htmldir3.t: 3rd test unit is now expected to fail.  So now the test file
demonstrates that the unit fails whether it's placed before the second
unit or after.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
_transform(): we only need one element from $self for this non-method function

Remove superseded variable and commented-out code

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Rename init_globals() to initialize()

Per suggestion from rjbs

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
ext/Pod-Html/t/*.t: Have each file run in tempdir

Until now, the tests for this library ran inside the ext/Pod-Html
directory in the core distribution.  Since the tests necessarily create
files while the harness runs, there was a possibility for race
conditions, which meant that we had to have a workaround in t/harness to
have the tests in this directory not run in parallel.

This commit restructures the testing process so that each test program
gets its own tempdir in which to create files and directories.

The test architecture is now found in t/lib/Testing.pm.
This module will export two functions on demand that are used in the
individual test files.  They guarantee that all files created when
running a test program will be created in a temporary directory rather
than underneath the core distribution.

t/pod2html-lib.pl is now superseded by t/lib/Testing.pm and can be
removed.  The workaround has been removed from t/harness.

Change directory name from testdir/ to corpus/.  So as to more precisely
describe what the directory's purpose is.

Keep porting tests happy. Update MANIFEST and known_pod_issues database.

Acknowledge Craig Berry's recommendation in earlier branch

This should help in the resolution of
https://github.com/Perl/perl5/issues/12271.

jkeenan committed:
use Pod::Simple::TiedOutFH rather than require-ing it

Anticipating CPAN release of https://github.com/perl-pod/pod-simple/pull/131

jkeenan committed:
Rename one method per suggestion from rjbs

jkeenan committed:
Explicitly import unixify() into Pod::Simple::XHTML::LocalPodLinks

Otherwise, you either have to fully qualify its invocation, or perl goes
searching for it in parent Pod::Simple::XHTML.

Prompted by query from rjbs.

jkeenan committed:
Delete commented-out code

oodler577 commented:
I putting the "conclusive" comment here for all of the PRs since they are cumulative - but the only thing that really concerns me is the non-isolated modification of `@ARGV`, which I saw in at least 2 places; i.e.,:

    unshift @ARGV, split ' ', $Config{pod2html} if $Config{pod2html};

You may be purposefully affecting this outside the scope of the subroutines, but there's also a chance this is not on purpose.

jkeenan committed:
Correct typo spotted by rjbs

jkeenan committed:
Pod-Html: rename Auxiliary.pm to Util.pm

Per rjbs code review in https://github.com/Perl/perl5/pull/18950

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
After eval-ing $VERSION, assign return value back to $VERSION

Remove superfluous assignment to @ISA.

Per rjbs code review in https://github.com/Perl/perl5/pull/18950

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Use lowercase for lexical variables -- even filehandles

Per rjbs code review in https://github.com/Perl/perl5/pull/18950

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Catch one more instance of 'Auxiliary'

jkeenan commented:
Closing this p.r., as it is now superseded by a direct p.r. into Perl5 blead, https://github.com/Perl/perl5/pull/19050.
